crates/apollo-compiler/src/validation/value.rs

```diff
@@ -0,0 +1,263 @@
+                        }
+                    }
+                }
+                _ => diagnostics.push(unsupported_type!(db, val, ty)),
+            },
+            // GraphQL has a constant literal to represent enum input values.
+            // GraphQL string literals must not be accepted as an enum input and
+            // instead raise a request error.
+            Value::Enum { ref value, loc } => match &type_def {
+                TypeDefinition::EnumTypeDefinition(enum_) => {
+                    let enum_val = enum_.values().find(|v| v.enum_value() == value.src());
+                    if enum_val.is_none() {
+                        diagnostics.push(
+                            ApolloDiagnostic::new(
+                                db,
+                                (*loc).into(),
+                                DiagnosticData::UndefinedValue {
+                                    value: value.src().into(),
+                                    definition: type_def.name().into(),
+                                },
+                            )
+                            .label(Label::new(
+                                val.loc(),
+                                format!("does not exist on `{}` type", type_def.name()),
+                            )),
+                        );
+                    }
+                }
+                _ => diagnostics.push(unsupported_type!(db, val, ty)),
+            },
+            // When expected as an input, list values are accepted only when
+            // each item in the list can be accepted by the list’s item type.
+            //
+            // If the value passed as an input to a list type is not a list and
+            // not the null value, then the result of input coercion is a list
+            // of size one, where the single item value is the result of input
+            // coercion for the list’s item type on the provided value (note
+            // this may apply recursively for nested lists).
+            Value::List { value: ref li, .. } => match &type_def {
+                TypeDefinition::ScalarTypeDefinition(_)
+                | TypeDefinition::EnumTypeDefinition(_)
+                | TypeDefinition::InputObjectTypeDefinition(_) => li
+                    .iter()
+                    .for_each(|v| value_of_correct_type(db, ty, v, var_defs.clone(), diagnostics)),
+                _ => diagnostics.push(unsupported_type!(db, val, ty)),
+            },
+            Value::Object { value: ref obj, .. } => match &type_def {
+                TypeDefinition::InputObjectTypeDefinition(input_obj) => {
+                    let undefined_field = obj.iter().find_map(|(name, value)| {
+                        let is_undefined = !input_obj.fields().any(|f| f.name() == name.src());
+                        if is_undefined {
+                            return Some((name, value));
+                        }
+                        None
+                    });
```

Comment:
```suggestion
                    let undefined_field = obj.iter().find(|(name, value)| {
                        !input_obj.fields().any(|f| f.name() == name.src())
                    });
```
I think this should work